### ISSUE TITLE
Support map values and nested values for attributes

### ISSUE THREAD

**Issue by tigrannajaryan** [open-telemetry/opentelemetry-specification#376]
After https://github.com/open-telemetry/opentelemetry-specification/pull/368 gets merged we will have support for array values.

If we add support for maps and nesting it will allow to represent arbitrary nested data structures in attribute values if needed.

This will apply to span and resource attributes.

**Comment by jmacd** [comment]
Can we explicitly state that this applies to resources too? I believe that span attributes and resources (which are only specified in the proto, currently) are specified with the same structure.

**Comment by Oberon00** [comment]
Are there any use cases for arbitrary nesting? I think (multi)maps would be useful to store, e.g., HTTP headers, but what would be the rationale for arbitrary nesting?

**Comment by shengxil** [comment]
Arbitrary nesting map can represent the classified values e.g. `{"http" : {"url":...,"method":...}}` or `{"sql" : {"query":...,"engine":...}}`. It can also host vendor specific data like `{"aws": {"account_id":...}}` in the situations when Resource isn't the right place, e.g. client side metrics

**Comment by jmacd** [comment]
In #579, Tigran's example seems to contain a use-case. The resource of "application B" is a set of key-value attributes.

**Comment by mwear** [comment]
Semantically, I agree that the dotted string notation is equivalent to a map, although I'd like to point out, that at least from the tracing client perspective, dotted strings are a slightly more efficient representation.

Consider the following representations for the key-value pair 'http.method': 'GET'.

**Dotted string representation**
`{ 'http.method': 'GET }`
To represent this we need 1 map and 2 strings; 3 total objects.

**Map representation**
`{ { 'http' : { 'method': 'get' } }`
This requires 2 maps, 3 strings; 5 total objects.


Furthermore, most tracing backends do not support nested attributes (as far as I know), and will need to flatten them into dotted strings. This is something that will either have to been done in the tracing clients during export, or by the backends on ingest.

While I recognize that this does have some advantages in regards to semantics and for the data that can be represented, it does introduce complexity into tracing clients and backends. I'm not saying we shouldn't do pursue this proposal, but we should actual benefits are, and whether the added complexity is worth the tradeoff.

**Comment by mwear** [comment]
I'd also like to add tracing clients can reduce the runtime string allocations to 0 for attribute keys by introducing constants for semantic conventions (and any other commonly used keys). We would lose this ability by changing to nested maps.

I should also clarify that I am completely ok with array support. It's the nested map support that I have reservations about.

**Comment by tigrannajaryan** [comment]
@bogdandrutu can you please clarify why is this reopened?

**Comment by bogdandrutu** [comment]
@tigrannajaryan because of the last week discussion and concerns raised by @mwear

**Comment by Oberon00** [comment]
Was the close even intentional? I can't remember any final decision in this issue. At least it's not documented here?

**Comment by Oberon00** [comment]
I guess closing this via a commit into some personal repository (tigrannajaryan/exp-otelproto@1507a2f) was accidental? It's surprising to me that GitHub allows this (I would even be tempted to call that a bug).

**Comment by tigrannajaryan** [comment]
> I guess closing this via a commit into some personal repository ([tigrannajaryan/exp-otelproto@1507a2f](https://github.com/tigrannajaryan/exp-otelproto/commit/1507a2f)) was accidental? It's surprising to me that GitHub allows this (I would even be tempted to call that a bug).

Definitely not intentional. I don't understand why does a commit in my personal repo which is not even a fork of this result in closing an issue here. That's weird. Agree it appears to be a bug.

**Comment by lizthegrey** [comment]
To clarify: are nested arrays allowed, e.g. [][]int, or only single dimensional arrays e.g. []int

**Comment by Oberon00** [comment]
Currently only homogeneous, single-dimensional arrays of primitives are allowed. This issue is about relaxing that.

**Comment by lizthegrey** [comment]
> Currently only homogeneous, single-dimensional arrays of primitives are allowed. This issue is about relaxing that.

Great. So my linked PR at the moment meets the spec, but won't once we relax the criteria.

**Comment by Oberon00** [comment]
I maintain we should not relax the criteria, but see #596. This has also been discussed in a few SIG Spec meetings.

**Comment by SergeyKanzhelev** [comment]
Corresponding PR didn't recieve a single approve. https://github.com/open-telemetry/opentelemetry-specification/pull/596#issuecomment-673883652 Meaning nobody is seeking for this feature. Moving this issue to Future milestone.

**Comment by tigrannajaryan** [comment]
For the record for future discussions: map values and nested values are currently supported at the OTLP protocol level, but are not utilized by the OpenTelemetry API. This issue is to discuss whether we also want to support such values in the API.

**Comment by bryannaegele** [comment]
Given map values and nested values are supported at the OTLP protocol level, should we treat the current spec of `MUST` for only primitives and homogeneous lists to be the base requirement and potentially add a `MAY` for maps and nested values? The expanded types have been working well at a trace level, so strictly implementing the spec would require us to add validations and restrict inputs, which I'm hesitant to add when it's working and we remove the need for added operations.

https://github.com/open-telemetry/opentelemetry-erlang/issues/111

**Comment by yurishkuro** [comment]
This issue is missing specific use cases / requirements. Here's one: capturing all http headers of a request.

**Comment by mateuszrzeszutek** [comment]
I have a couple of questions/observations regarding the HTTP headers use case:

1. If we allow adding nested structures/maps to `Attributes`, what type should the attribute key represent? Also `Attributes`? E.g. in Java `AttributeKey<Attributes> attributesKey(String key)` would be the factory method
2. How do we serialize nested structures in non-OTLP formats? Currently in Jaeger & Zipkin exporters we're serializing arrays as JSON lists (`["a","b","c"]`), by analogy should we serialize nested structures as JSON objects? Or should flatten them like describes in this comment https://github.com/open-telemetry/opentelemetry-specification/issues/376#issuecomment-643488960 ?
3. Now that `Attributes` are also used in the metrics API, what would happen if you passed a nested `Attributes` instance to a meter? Since [they're not allowed for metrics](https://github.com/open-telemetry/opentelemetry-specification/issues/1061#issuecomment-908480390)

**Comment by bogdandrutu** [comment]
>If we allow adding nested structures/maps to Attributes, what type should the attribute key represent? Also Attributes? E.g. in Java AttributeKey<Attributes> attributesKey(String key) would be the factory method

That is a very Java specific problem and the Java SIG should reach consensus.

> How do we serialize nested structures in non-OTLP formats? Currently in Jaeger & Zipkin exporters we're serializing arrays as JSON lists (["a","b","c"]), by analogy should we serialize nested structures as JSON objects? Or should flatten them like describes in this comment Support map values and nested values for attributes #376 (comment) ?

JSON is the simplest way. I think we should not enter the business of flattening. Also each protocol can define their ways of doing this, we care mostly about Jaeger and Zipkin.

> Now that Attributes are also used in the metrics API, what would happen if you passed a nested Attributes instance to a meter? Since they're not allowed for metrics

They will continue to not be allowed for the moment

**Comment by Oberon00** [comment]
>> If we allow adding nested structures/maps to Attributes, what type should the attribute key represent? Also Attributes? E.g. in Java AttributeKey attributesKey(String key) would be the factory method
>
> That is a very Java specific problem and the Java SIG should reach consensus.

I don't think this is Java-specific at all. This is a problem for all SIGs that want to strongly type their attributes and also for the semantic convention generator. What should the type of the attribute be declared as? a `map<string, string>` do we need arbitrary nesting or just maps from atomic primitive to atomic primitive or array thereof?

I think we need to understand that while allowing maps seems conceptually very simple, and also a no-op in OTLP, I think it opens a can of worms regarding tooling.

Also, I know of no backend that supports maps for tracing. @mwear also brought up performance concerns in above in https://github.com/open-telemetry/opentelemetry-specification/issues/376#issuecomment-643488960.

**Comment by mateuszrzeszutek** [comment]
> I don't think this is Java-specific at all. This is a problem for all SIGs that want to strongly type their attributes and also for the semantic convention generator. What should the type of the attribute be declared as? A `map<string, string>`? That's our first complex type (arrays are currently coded by explicitly listing all possible array types `string[]`, `double[]` etc.). Do we need arbitrary nesting or just maps from atomic primitive to atomic primitive or array thereof?

Exactly - for the HTTP headers sue case the desired type would be `map<string, string[]>`, but what about other uses? Should we have to add `map<string, string>` too? Or `map<string, ?>`, `map<int, string>` etc. Or just say that we support nesting `Attributes` (which is sort of a map) inside `Attributes`. 

> I think we need to understand that while allowing maps seems conceptually very simple, and also a no-op in OTLP, I think it opens a can of worms regarding tooling.
> 
> Also, I know of no backend that supports maps for tracing. @mwear also brought up performance concerns above in [#376 (comment)](https://github.com/open-telemetry/opentelemetry-specification/issues/376#issuecomment-643488960) (and received many upvotes on that comment).

So the question here is: is it really worthwhile to introduce map-valued/nested attributes? Do we have any other use case than HTTP attributes? Because if it's only that one use case the cost of implementing that seems to be pretty high (new attribute key  type support + exporter changes in all language SIGs, collector included; backend support).

**Comment by tigrannajaryan** [comment]
FYI, OTLP and Collector define the `attributes` field as `map<string,AnyValue>`, where `AnyValue` is one of the supported value types, including a map, so it is a recursive definition.

>Because if it's only that one use case the cost of implementing that seems to be pretty high (new attribute key type support + exporter changes in all language SIGs, collector included; backend support).

No changes are needed in the Collector, it is already fully supported.

**Comment by mateuszrzeszutek** [comment]
> No changes are needed in the Collector, it is already fully supported.

**Comment by Oberon00** [comment]
> No changes are needed in the Collector, it is already fully supported.

How does the collector handling a nested map to Jaeger / Zipkin? Does it already JSON-encode?

**Comment by tigrannajaryan** [comment]
> Also, I know of no backend that supports maps for tracing.

From what I understand AS X-Ray does: https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-annotations

From the above link:

>Metadata are key-value pairs with values of any type, including objects and lists

**Comment by tigrannajaryan** [comment]
> > No changes are needed in the Collector, it is already fully supported.
> 
> How does the collector handle exporting a nested map to Jaeger / Zipkin? Does it already JSON-encode?

Yes. Jaeger translation [calls AsString()](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0d78df6b0455fd685cf6d9dcb654986e7de7ac86/pkg/translator/jaeger/traces_to_jaegerproto.go#L160) which [converts maps and arrays](https://github.com/open-telemetry/opentelemetry-collector/blob/27d80d1391617cf20b990417eaf5e793c9bc74bd/model/pdata/common.go#L401) to JSON string.

I didn't check Zipkin, it should do the same.

**Comment by mateuszrzeszutek** [comment]
> > Also, I know of no backend that supports maps for tracing.
> 
> From what I understand AS X-Ray does: https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-annotations
> 
> From the above link:
> 
> > Metadata are key-value pairs with values of any type, including objects and lists

It's worth noting that AWS Metadata are not indexed/searchable:

> Metadata are key-value pairs with values of any type, including objects and lists, but that are not indexed. Use metadata to record data you want to store in the trace but don't need to use for searching traces.

If you don't want to search by specific metadata values and just treat the whole object as one, indivisible entity then you might as well just serialize them to JSON and put them all into a string attribute; there's no real need to have a nested structure of attributes for them.

**Comment by tigrannajaryan** [comment]
I can see a few exporters (in addition to X-ray) in the Collector which treat the nested like maps instead of flattening or converting to JSON or strings, so supposedly their corresponding backends support nested structures. I think this issue requires more feedback from vendors before we decide what to do about it.

**Comment by scheler** [comment]
> So the question here is: is it really worthwhile to introduce map-valued/nested attributes? Do we have any other use case than HTTP attributes? Because if it's only that one use case the cost of implementing that seems to be pretty high (new attribute key type support + exporter changes in all language SIGs, collector included; backend support).

Is there an alternative suggested to represent map data in attributes? One of the instrumentation libraries in JS sdk for browser sends map data as Span Events which is very inefficient.


```
{
    "traceId": "bb2a647c845bb6d68dc415042f0ee49d",
    "name": "documentLoad",
    "id": "69a3580422564f5c",
    "kind": 0,
    "timestamp": 1654107849066400,
    "duration": 121600,
    "attributes": {
        "component": "document-load",
        "http.url": "http://localhost:5002/",
        "http.user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36",
 

    },
    "status": {
        "code": 0
    },
    "events": [

      {

}

        {
            "name": "fetchStart",
            "attributes": {},
            "time": [
                1654107849,
                66400100.00000001
            ]
        },
        {
            "name": "unloadEventStart",
            "attributes": {},
            "time": [
                1654107849,
                94600100
            ]
        },
        {
            "name": "unloadEventEnd",
            "attributes": {},
            "time": [
                1654107849,
                97000100
            ]
        },
        {
            "name": "domInteractive",
            "attributes": {},
            "time": [
                1654107849,
                104000100
            ]
        },
        {
            "name": "domContentLoadedEventStart",
            "attributes": {},
            "time": [
                1654107849,
                187800100
            ]
        },
        {
            "name": "domContentLoadedEventEnd",
            "attributes": {},
            "time": [
                1654107849,
                187900100
            ]
        },
        {
            "name": "domComplete",
            "attributes": {},
            "time": [
                1654107849,
                188000100
            ]
        },
        {
            "name": "loadEventStart",
            "attributes": {},
            "time": [
                1654107849,
                188000100
            ]
        },
        {
            "name": "loadEventEnd",
            "attributes": {},
            "time": [
                1654107849,
                188000100
            ]
        },
        {
            "name": "firstPaint",
            "attributes": {},
            "time": [
                1654107849,
                106000100
            ]
        },
        {
            "name": "firstContentfulPaint",
            "attributes": {},
            "time": [
                1654107849,
                106000100
            ]
        }
    ]
}

```

**Comment by Oberon00** [comment]
The alternative is to use dot notation in your key names so instead of nesting `foo` under `bar`, you use `bar.foo` as key.
I don't see any nested attributes in the example JSON you posted.

**Comment by scheler** [comment]
> This is actually used in the semantic conventions for HTTP headers

This is not generalized though - there is no clarity on when the receivers should convert the dotted keys to maps. A reserved keyword, say, `_map_` would have been better. so when you see `http.request.headers._map_.<key>`, you know `http.request.headers` is a map. Is it possible to have reserved keywords in attribute name components?

**Comment by Oberon00** [comment]
>  there is no clarity on when the receivers should convert the dotted keys to maps.

What would be the use case for converting to a map on the receiver? If you need a map for a particular use case, then you know that particular data, e.g. in a particular namespace, is to be interpreted as a map. If you don't know how to interpret the data, what benefit would you have from a map?

> For the JSON I posted, it would look like this if there was support for nested attributes

This looks  like you you want to have events instead of nested attributes, since events are exactly meant for marking points in time. For "unload" you may also consider a (child) span, since it has start & end.

**Comment by scheler** [comment]
> What would be the use case for converting to a map on the receiver?

Having a 2-way conversion provides way for future interoperability. It would be helpful if the flattened data needs to be converted back to maps.

I support map type for attribute values across resources, spans and logs, flattening them during export wherever required using a certain reversible convention. 

Regarding @mwear 's argument on additional string allocations with maps, note that it does not apply to the 2 use-cases described here so far - for `http.request.header.<key>` and for the `timingData` example above, the keys within the map are static. Although the concern is valid in general, I think there's a bigger benefit in having a smaller payload on the wire.

**Comment by Oberon00** [comment]
If you want smaller payload, you should first enable transport compression, then minimize the number of spans you create, the number of attributes you set on the span. I don't think this is a good enough argument for map values.

**Comment by scheler** [comment]
@Oberon00 would that be the counter argument for the ask here?

**Comment by scheler** [comment]
Smaller payload is a side benefit. The primary motivation for this ask is that the data is generated as a map, and the receivers need them as maps. So, the ask for map support for attribute values is to avoid unnecessary flattening during transport.

**Comment by tsloughter** [comment]
What is the status of this? Is it just in need of someone to write the spec PR?

**Comment by tigrannajaryan** [comment]
> What is the status of this? Is it just in need of someone to write the spec PR?

Not everyone in the community believes it is necessary (see above). So, it is not just to write a PR, but also to find arguments that will convince people.

**Comment by tsloughter** [comment]
Ah ok, I guess I need to take a closer look.

The main "objection" I saw seemed to be more of an alternative implementation and not an objection to map values in the API, but in the protocol -- i.e. the dot-notation based approach, which could be a translation of the user level map used in the API to the proto representation.

**Comment by cartermp** [comment]
(this was also posted in slack)

Dunno if this helps regarding motivation, but:

In my work, we're getting customers who are attaching objects as json-encoded strings as attributes on spans. They would like our backend to "splat" that object into fields in the backend so that they can query for the presence of a value in one of the fields in that object. However, without a way to know for sure that the string is actually a JSON-encoded object, telemetry backends need to scan every string in every attribute and run JSON validation, which is extremely expensive. The same would apply to any span processor in the collector.

The current thing we're telling customers to do is to see if they can live without it. That doesn't appear to work. Another is to ask them to manually encode an object into each span. That's really difficult for larger objects, or a lot of smaller objects. We could consider a collector processor that allows them to configure keys that they know up-front represent objects, and then unpack only those (or do the same in a telemetry backend). But that's also not ideal, because developers in a large organization may not have access to the collector config they're using, and it's just additional work for them to have to coordinate collector config changes with instrumentation they write in the app (or backend changes in a tool). Another option is to switch your language to JavaScript, since `addAttributes` takes advantage of the fact that JS objects are key-value pairs that can be represented as a set of attributes (to an extent). But clearly, asking someone to use a different language isn't a good solution.

In the interim, having an endorsed way around this problem that people can follow (and not just my customers) would be great to know so that we can document it in the docs.

**Comment by tsloughter** [comment]
To build on what @cartermp is saying, I also think the user's need for querying is why it should be typed and not just string values in the protos. Maybe no system has this ability today but knowing the type is an integer of some attribute within a map means both the backend can more efficiently store and search that value but the user can be provided more ways of querying the data (without having to wrap their query reference to the attribute value in `atoi`).

**Comment by tigrannajaryan** [comment]
>In my work, we're getting customers who are attaching objects as json-encoded strings as attributes on spans. They would like our backend to "splat" that object into fields in the backend so that they can query for the presence of a value in one of the fields in that object.

One alternate way that this particular use case can be addressed is by allowing to record the fact that a string is a JSON in the attribute value. This would allow the backends to know when to unparse it as a JSON. There are downsides to this though since we require everyone who wants to read/modify any attribute (e.g. the Otel Collector) to have special processing logic that deserializes JSON, reads/modifies, then serializes again. It may result in awkward user experience where operations on values inside JSON-encoded fields have their own syntax to use in the Otel Collector processors. I am not sure I like this approach, but wanted to list it for completeness. IMO, first-class support for non-primitive types is a more flexible and powerful alternate.

**Comment by Oberon00** [comment]
> The main "objection" I saw seemed to be more of an alternative implementation and not an objection to map values in the API,

My main objection is that it makes the whole data model more complicated, from the APIs to the backend processing, as well as the semantic conventions (should I use `.` or a subobject?) and the semantic convention tooling (what syntax do we use for sub-objects, since the dot is already used in attribute names?)

**Comment by tsloughter** [comment]
Ok, so I misunderstood and it is a full on objection to having nested values in attributes?

Are there cases where use of `.` in the semantic conventions would change the meaning if it was considered nesting instead of a single name? My thinking was all those `.`'s in conventions already are basically representing nesting, so there would be no conflict.

**Comment by Oberon00** [comment]
>  My thinking was all those .'s in conventions already are basically representing nesting, so there would be no conflict.

So are you suggesting that after supporting nested attributes, every every semantic convention (that uses a dot, which is every single one) will be sent as a nested attribute? What about "service.name" which was declared stable already?

**Comment by tsloughter** [comment]
I'm suggesting they would be considered to be nested. `service.name` and `service.instance.id` are attributes  under `service`. And upgrade to the protocol would introduce actual maps but would have to support the `.` form as well. OTLP servers would have to support both to not require a major version bump and continue to support the stable form `service.name`. 

I don't work on the collector so am only voicing a possibility, the work certainly could be way more than the potential advantages --  many deeply nested attributes when keeping the dot notation as the only option would be duplicating the strings a lot.

But just stating the dot notation represents nested would allow backends to know they can split up these attributes and still get typed values for each nested attribute when they index and search. Just means accepting the memory duplication you wouldn't have with shoving in a json blob.

**Comment by Oberon00** [comment]
>  OTLP servers would have to support both to not require a major version bump 

If we require OTLP servers to support something new, that's a major version bump.

**Comment by tsloughter** [comment]
True, I was mainly thinking about the client not wanting to have a major bump, but you are right, it would be a major bump, so I guess off the table. 

In which case only supporting the `.` notation as nesting or adding maps as values would work that don't involve telling users to shove json blobs in -- which I'm not a fan of because it means a user has to do `data -> json -> pb` and include a json library.

**Comment by tigrannajaryan** [comment]
Going from dot-notation of attribute _names_ to hierarchical is a different topic. If you would like to discuss it please open a separate issue to avoid confusion. 

This issue is about allowing attribute _values_ which are maps or other nested values. For example I may want to record the following information about a TCP connection:

```json
{
  "source.endpoint": { "host": "client.com", "port": 1234 }
  "destination.endpoint": { "host": "server.com", "port": 2345 }
}
```

Note that each attribute's value is a map.

**Comment by Oberon00** [comment]
I guess this discussion branched up from my above comment:

> the semantic convention tooling (what syntax do we use for sub-objects, since the dot is already used in attribute names?)

I.e. how would you define a semantic convention that you may have an object with `host` and `port` properties under an attribute name `source.endpoint`, and how would you denote the full "path" of the host property here (maybe with `source.endpoint#host`, borrowing from Javadoc), or are we sure we'd never want that (probably not)?

**Comment by tsloughter** [comment]
Right, I think it is the same topic (as in, instead of adding maps, declare `source.endpoint.host` and `source.endpoint.port` can/should be parsed into maps like your example by a backend, bc otherwise I saw what looked like a deadend, hope I'm wrong and maps get accepted). But I'll will cede to @tigrannajaryan saying it is and not continue that discussion on this thread :)

**Comment by tsloughter** [comment]
I do want to once more iterate that from a user perspective not having maps as values is going to hurt the otel experience because different vendors will offer different forms of nested attributes. There should be on doubt that json is going to get used in values and eventually parsed by some vendors into attributes. I don't want to see that and think the issues raised so far aren't blockers but issues to be resolved in the PR that adds support for maps and updates the semantic conventions.

**Comment by Oberon00** [comment]
> Right, I think it is the same topic (as in, instead of adding maps, declare source.endpoint.host and source.endpoint.port can/should be parsed into maps

So then the example posted by @tigrannajaryan with the mixture of dots in names and nested objects in https://github.com/open-telemetry/opentelemetry-specification/issues/376#issuecomment-1230397133 would be something that you do not want?

**Comment by Oberon00** [comment]
> I do want to once more iterate that from a user perspective not having maps as values is going to hurt the otel experience because different vendors will offer different forms of nested attributes

If there is pressure from vendors to support this, ideally with prior art, that would be a good argument to support this IMHO. Any examples?

**Comment by cartermp** [comment]
> If there is pressure from vendors to support this, ideally with prior art, that would be a good argument to support this IMHO. Any examples?

This is how Honeycomb works for non-OTLP data FWIW. It's been a feature since long before OpenTelemetry, for all event and tracing data sources. Another example that I know does this today is [Seq](https://datalust.co/seq).

**Comment by Oberon00** [comment]
It seems fine to me to introduce it as a feature for backwards-compatibility with vendors that support this. But using it in our semantic conventions could make people want to re-design them alltogether (as @tsloughter mentioned one might want to replace dotted names with subobjects), and that decision should be synced with the semantic conventions SIG @jsuereth. If we decide to use nesting in our semantic conventions generally, IMHO we should also give that idea & implementations thereof more time to "bake" before we declare anything stable.

**Comment by Oberon00** [comment]
So I think there are more issues here that need decisions:
* Technical support for nested values in SDKs, APIs, OTLP exporters, collector, OTLP backends, UIs (needs to be added in a backwards-compatible way which might not be trivial everywhere but otherwise
* Decision to use them or not in any new semantic conventions in the future. If so, requires support in the semantic convention generator/YAML, schema files, in any languages semantic convention artifacts (some use strong typing for keys, e.g. generated Java code https://github.com/open-telemetry/opentelemetry-java/blob/main/semconv/src/main/java/io/opentelemetry/semconv/trace/attributes/SemanticAttributes.java)
* Decision to use them or not in existing semantic conventions. If so, may require further changes to backends that interpret the semantic conventions in some way. Especially service.name seems important.

**Comment by tigrannajaryan** [comment]
One more use-case were map values would be beneficial: https://github.com/open-telemetry/opentelemetry-specification/pull/2841

**Comment by tigrannajaryan** [comment]
Another use-case here: https://github.com/open-telemetry/oteps/pull/219

**Comment by tsloughter** [comment]
#2841 is an example where the complexity for both sides (client and backend) is increased by lack of map support. The semantic convention is also harder to understand when it has to be split into pieces where it could be that the whole, like `exception.structured_stacktrace` is optional but within the whole there may be those that are required in `structured_stacktrace` but still are marked as optional since they technically are optional in the attributes since there may be no `structured_stacktrace`.

**Comment by jmacd** [comment]
I support map-valued attributes. On a technical level, this will not present a significant problem for the OTel-Go SDK based on my understanding of the issues. We already admit slice-valued attributes [and although there was a bug in this support, it was recently fixed](https://github.com/open-telemetry/opentelemetry-go/pull/3252), and we already have a data type (`attribute.Set`) which represents a unique attribute-value map. To add a Map-valued attribute, we would add the `attribute.Set` as one of the possible Value types. (Likewise, to add List-of-attributes-valued attributes, we would apply the technique used for primitive-valued slices.)

@Oberon00 list of what it takes to finish this support looks good: https://github.com/open-telemetry/opentelemetry-specification/issues/376#issuecomment-1232783131

**Comment by yurishkuro** [comment]
I also support maps as attribute values.

I don't completely agree with @Oberon00's list https://github.com/open-telemetry/opentelemetry-specification/issues/376#issuecomment-1232783131, I think it broadens the scope of this issue more than necessary and makes it difficult to make progress. Specifically, I am talking about the two points on deciding whether maps should be used in future or existing semantic conventions. Yes, those are important questions, but they do not need to block the API change itself, since there are use cases for map attributes that are not about semantic conventions, but about capturing app-specific data.

**Comment by tigrannajaryan** [comment]
Here is a draft PR to discuss: https://github.com/open-telemetry/opentelemetry-specification/pull/2888

**Comment by austinlparker** [comment]
Closed by #3858

**Comment by adriangb** [comment]
I have to say I'm quite disappointed with this outcome.

IMO it's not a huge ask for backends to flatten nested attributes to dot notation if they don't want to support them, at least compared to https://github.com/open-telemetry/opentelemetry-specification/pull/4333 which is a much bigger change for backends and where my comment of "but this will make it harder for backends" was dismissed with "that sounds like a backend problem". Concretely a good chunk of (the majority of? I don't have hard numbers) new backends seem to be built on top of ClickHouse, which has JSON support. Other modern open source query engines like DuckDB, DataFusion all have JSON support. I can't help but feel that a lot of decisions in OpenTelemetry are driven by the needs of existing vendors.

It's also *extremely* confusing the differences in what is supported and not at the SDK level vs. the spec level vs. the protocol level, e.g. see https://github.com/open-telemetry/opentelemetry-rust/pull/2471#discussion_r1941628304

**Comment by pellared** [comment]
@adriangb, You can also check this https://github.com/open-telemetry/opentelemetry-specification/pull/2888

> I can't help but feel that a lot of decisions in OpenTelemetry are driven by the needs of existing vendors.

I want just to call out that some vendors where supporting the change (extending the standard attributes) so please do not but everyone to "existing vendors" bucket.

For instance, as far as I remember all @open-telemetry/go-maintainers (Google, Elastic, Cisco, Splunk) were in favor of this change. I think we lost our will after a lot of feedback which was against the change.

Given https://github.com/open-telemetry/opentelemetry-specification/pull/3858  I am not sure if there is a room for changing the decision. @open-telemetry/technical-committee, opinions?

**Comment by pellared** [comment]
As far as I understand, the OTel Collector does support map values and nested values for attributes in all signals. For reference:

- https://pkg.go.dev/go.opentelemetry.io/collector/pdata/ptrace#Span.Attributes
- https://pkg.go.dev/go.opentelemetry.io/collector/pdata/pmetric#NumberDataPoint.Attributes
- https://pkg.go.dev/go.opentelemetry.io/collector/pdata/plog#LogRecord.Attributes
- https://pkg.go.dev/go.opentelemetry.io/collector/pdata/pcommon#Map

@mx-psi, am I correct ? If so then are you experiencing any issues connected with it? If not then maybe we should really revisit it and do it in all languages? According to the specification it would be a bug (specification incompliance) in the collector. If the collector would decide to keep the current design, then I do not see any reason why all languages would have to be different.

**Comment by mx-psi** [comment]
> As far as I see, the OTel Collector does support heterogeneous map and array values, nested values, bytes for attributes in all signals (including metrics, right?). For reference:
> 
> * https://pkg.go.dev/go.opentelemetry.io/collector/pdata/ptrace#Span.Attributes
> * https://pkg.go.dev/go.opentelemetry.io/collector/pdata/pmetric#NumberDataPoint.Attributes
> * https://pkg.go.dev/go.opentelemetry.io/collector/pdata/plog#LogRecord.Attributes
> * https://pkg.go.dev/go.opentelemetry.io/collector/pdata/pcommon#Map
> 
> @mx-psi, am I correct? If so then are you experiencing any issues connected with it? If not then maybe we should really revisit it and do it in all languages? 

I am not aware of any issues, but I think we should ask the component owners using this. Here are some Github searches that show usages in the wild of `pcommon.Map` related functions. Note that there are uses of this type that are not intended to construct 'complex attributes', in particular the [log body](https://pkg.go.dev/go.opentelemetry.io/collector/pdata/plog#LogRecord.Body) may be a map per the [data model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body):

- [Uses of `pcommon.NewValueMap` outside of tests](https://github.com/search?q=language%3AGo+pcommon.NewValueMap+-path%3A%2F_test.go%2F+-is%3Afork+-repo%3Apdelewski%2Fopentelemetry-collector-contrib&type=code) (11 hits, most of them come from two components)
- [Uses of `.PutEmptyMap` outside of tests](https://github.com/search?q=language%3AGo+%2F%5C.PutEmptyMap%2F+-path%3A%2F_test.go%2F+-is%3Afork&type=code&p=2)
- [Uses of `pcommon.ValueTypeMap` outside of tests](https://github.com/search?q=language%3AGo+%2Fpcommon%5C.ValueTypeMap%2F+-path%3A%2F_test.go%2F+-is%3Afork&type=code) (I would expect all exporters to use this one to convert to their own format)

The only interesting use case I have found is [parsing XML in attributes in OTTL](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6638b83fad429c9636ac3e40d00c2b3acafebdc1/pkg/ottl/ottlfuncs/func_parse_xml.go#L118). Maybe @TylerHelmuth  @evan-bradley can talk about their experience here?

> According to the specification it would be a bug (specification incompliance) in the collector. If the collector would decide to keep the current design, then I do not understand why all languages (including OTel Go) would have to be different.

I believe this is because the proto actually supports these: https://github.com/open-telemetry/opentelemetry-proto/blob/6a4f4e4fd867eda646ae14604dbc773e2516826a/opentelemetry/proto/common/v1/common.proto#L36-L37, and pdata is mostly a wrapper of the proto. If the Collector is not spec-compliant, is the proto also not specification compliant then?

**Comment by adriangb** [comment]
Thanks @pellared. I'm sorry if I was harsh to all here.

One more thing I want to point out is that some things are moving to use the Logs API from tracing in part because of this limitation: in https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-events/#tools nested Log bodies are being logged, and I suspect that Logs are being used because this data could not be sent via a span (this is a suspicion, I did not spend the time to dredge through the history of this decision). It's a shame to move things to a system that was initially designed for backwards compatibility with legacy systems just because of artificial limitations with span attributes.

**Comment by pellared** [comment]
> I suspect that Logs are being used because this data could not be sent via a span

I think that:
- not every event is inside a span/trace.
- it may be also about delivering events before the span ends.

**Comment by adriangb** [comment]
> it may be also about delivering events before the span ends (probably this could be solved differently)

Totally agreed, I think this is an issue. I think I mentioned this before but we've created a concept of "pending span" which is a span we send out with 0 duration as soon as a span is opened so that we can show long lived spans before they are closed. We would be very happy for there to be some sort of convention along the lines of "send a log / event with X data to notify a backend that a span has started".

> not every event is inside a span/trace

This I'm a bit confused by. I understand that you may not have a span currently running at the point when you want to emit an event, but what is stopping you from opening a span just to do that? Here's what we do:

```python
import logfire

logfire.info('an event/log')
```

Which behind the scenes we transform into (more complicated, we hook in deeper to mark this as a "log span", make duration 0 and avoid overheads):

```python
span = trace.start_current_span('an event/log')
span.end()
```

But that allows emitting events without a span currently active and without a new type of signal. Am I missing your point?

If anything this reinforces my point: if we allowed span attributes the full flexibility that log attributes we could go in the opposite direction and merge events into spans instead of merging events into logs (a signal initially designed for compatibility with legacy systems).

**Comment by pellared** [comment]
@adriangb, can you summarize and leave a comment here please: https://github.com/open-telemetry/opentelemetry-specification/issues/4393?

**Comment by adriangb** [comment]
> It is possible but it comes with unnecessary performance overhead and data noise. I see it as an ugly/hacky workaround and not a proper solution.

I don't think it's that much overhead or hack. It still goes out as a single thing over the wire. Users don't know the difference. It's not hard for backends to handle. And if there was an agreement on going in that direction it would be trivial to add the right APIs to the SDKs to do this with zero overhead and zero ambiguity.

**Comment by adriangb** [comment]
> [@adriangb](https://github.com/adriangb), can you summarize and leave a comment here please: [#4393](https://github.com/open-telemetry/opentelemetry-specification/issues/4393)?

[Done, thanks](https://github.com/open-telemetry/opentelemetry-specification/issues/4393#issuecomment-2640512576)

**Comment by TylerHelmuth** [comment]
@mx-psi OTTL has a plethora of ways to assign map values to attributes for any signal. We allow this because pdata allows it.

**Comment by pellared** [comment]
For transparency, we are getting there: https://opentelemetry.io/blog/2025/complex-attribute-types/

This is already supported by OTLP: https://github.com/open-telemetry/opentelemetry-proto/releases/tag/v1.9.0
